Fix heap_allocated reinsertion for refzero finalizer rescued objects

svaarala opened the pull request:
Fix heap_allocated reinsertion bug which affected refzero finalizer rescued objects. See test case in: https://github.com/oleavr/dukrecycle/ (a bug test case will be adapted in this branch based on that).

- [x] Fix the re-queueing of refzero finalizer rescued objects
- [x] Adapt test case from dukrecycle, test that bug appears before and gets fixes by the diff
- [x] Add linked list consistency assertion helper which checks that prev/next and curr are linked correctly
- [x] Add more heap_allocated consistency asserts to linked list manipulation points, ensure assertion tests pass with them in place and that asserts would have caught the original bug
- [x] Releases entry
- [x] Schedule backports to 1.4.1, 1.3.3, and 1.2.6

svaarala commented:
@joshengebretson If the diagnosis/fix here is correct, this may be related to the finalizer issues you had a way back :) Here's hoping!

oleavr commented:
This fixes the stability issues we were seeing in the [frida-gum](https://github.com/frida/frida-gum) test-suite! :heart:

svaarala commented:
@oleavr Big thanks for coming up with the repro case!

fatcerberus commented:
@svaarala What is the bug here, exactly?  I'm curious because I haven't yet seen the finalizer issues Josh had with Atomic in minisphere.

svaarala commented:
The diff really says it all, but in short, object rescued during refcount-triggered finalization need to be queued into the heap_allocated list. That queueing operation failed to update the "prev" pointer of the previous list head which resulted in an inconsistent heap_allocated list. With a suitable mix of later operations (e.g. finalizing again) this could trigger a read/write-after-free situation. It's a pretty significant bug and could have various weird side effects -- but only if an object is rescued during refcount-triggered finalization.

fatcerberus commented:
Could that potentially lead to a segfault?

svaarala commented:
I think so, though the problems were captured using ASAN/valgrind in practice.

svaarala commented:
Added the test case and checked that the test case catches the original issue.

Also added linked list consistency asserts where `duk_heaphdr` linked list is manipulated. Verified that test cases pass with the asserts in place, and that the asserts would have caught the original bug.
